java/ql/src/experimental/Security/CWE/CWE-522/UnsecureBasicAuth.ql

```diff
@@ -0,0 +1,213 @@
+            .(VarAccess)
+            .getVariable()
+            .getAnAssignedValue()
+            .(StringLiteral)
+            .getRepresentedString()
+            .matches("Basic %") or
+        arg1.(CompileTimeConstantExpr).getStringValue().matches("Basic %") or //post.addHeader("Authorization", "Basic ....")
+        arg1.(AddExpr).getLeftOperand().(StringLiteral).getRepresentedString().matches("Basic %") //post.addHeader("Authorization", "Basic "+authStringEnc)
+      )
+    ) and
+    exists(VarAccess request, VariableAssign va, ConstructorCall cc, Expr arg0 |
+      this.getQualifier() = request and //Check the method invocation with the pattern post.addHeader("Authorization", "Basic " + authStringEnc)
+      va.getDestVar() = request.getVariable() and
+      va.getSource() = cc and
+      cc.getArgument(0) = arg0 and
+      builtFromHttpStringConcat(arg0) // Check url string
+    )
+  }
+}
+
+/** The `openConnection` method of Java URL. Not to include `openStream` since it won't be used in this query. */
+class HttpURLOpenMethod extends Method {
+  HttpURLOpenMethod() {
+    this.getDeclaringType() instanceof TypeUrl and
+    this.getName() = "openConnection"
+  }
+}
+
+/**
+ * Tracks the flow of data from parameter of URL constructor to the url instance
+ */
+class URLConstructorTaintStep extends TaintTracking::AdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(URLConstructor u |
+      node1.asExpr() = u.stringArg() and
+      node2.asExpr() = u
+    )
+  }
+}
+
+/**
+ * Tracks the flow of data from `openConnection` method to the connection object
+ */
+class OpenHttpURLTaintStep extends TaintTracking::AdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(MethodAccess ma, VariableAssign va |
+      ma.getMethod() instanceof HttpURLOpenMethod and
+      ma.getQualifier() = node1.asExpr() and
+      (
+        ma = va.getSource()
+        or
+        exists(CastExpr ce |
+          ce.getExpr() = ma and
+          ce = va.getSource() and
+          ce.getControlFlowNode().getASuccessor() = va //With a type cast like HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+        )
+      ) and
+      node2.asExpr() = va.getDestVar().getAnAccess()
+    )
+  }
+}
+
+class HttpStringToHttpURLOpenMethodFlowConfig extends TaintTracking::Configuration {
+  HttpStringToHttpURLOpenMethodFlowConfig() {
+    this = "UnsecureBasicAuth::HttpStringToHttpURLOpenMethodFlowConfig"
+  }
+
+  override predicate isSource(DataFlow::Node src) { src.asExpr() instanceof HttpString }
+
+  override predicate isSink(DataFlow::Node sink) {
+    sink.asExpr().(VarAccess).getVariable().getType() instanceof TypeUrlConnection or
+    sink.asExpr().(VarAccess).getVariable().getType() instanceof TypeHttpUrlConnection // Somehow TypeHttpUrlConnection isn't instance of TypeUrlConnection
```

Comment:
The reason why they are not subclasses is likely because QL does and can not model the type hierarchy (as far as I know) since both have different names.

Reply:
Thanks for the explanation. I've removed the word "Somehow" from the comment.